From 46a9125f8f74e3d56ef76e7b0e4d502a95afbbc1 Mon Sep 17 00:00:00 2001
From: ValdikSS <iam@valdikss.org.ru>
Date: Tue, 20 Jan 2026 07:16:04 +0300
Subject: [PATCH 1/2] Add Dual Channel into Bluetooth Audio Channel Mode
 developer options menu

Add UI support for SBC Dual Channel (SBC HD) mode in the Bluetooth Audio
Channel Mode dialog within Developer Options.

This commit:
- Adds "Dual Channel" radio button to the channel mode dialog layout
- Updates BluetoothChannelModeDialogPreference to dynamically add the option
- Updates BluetoothChannelModeDialogPreferenceController to handle the new mode
- Updates AbstractBluetoothDialogPreferenceController with DUAL_CHANNEL constant
- Force-enables Dual Channel selection for SBC codec

Adapted for Android 16 QPR1 from LineageOS patches.

Co-authored-by: Flopster101 <nahuelgomez329@gmail.com>
Signed-off-by: Flopster101 <nahuelgomez329@gmail.com>
Change-Id: Icd3179ac63310355547fd58060e75f25b4a8210d
---
 .../bluetooth_audio_channel_mode_dialog.xml   |  4 ++++
 ...ctBluetoothDialogPreferenceController.java |  7 +++++-
 .../BluetoothChannelModeDialogPreference.java | 14 +++++++++---
 ...ChannelModeDialogPreferenceController.java | 22 +++++++++++++++++++
 4 files changed, 43 insertions(+), 4 deletions(-)

diff --git a/res/layout/bluetooth_audio_channel_mode_dialog.xml b/res/layout/bluetooth_audio_channel_mode_dialog.xml
index 9154ee4..ceba79c 100644
--- a/res/layout/bluetooth_audio_channel_mode_dialog.xml
+++ b/res/layout/bluetooth_audio_channel_mode_dialog.xml
@@ -42,6 +42,10 @@
             <include
                 android:id="@+id/bluetooth_audio_channel_mode_stereo"
                 layout="@layout/preference_widget_dialog_radiobutton"/>
+
+            <include
+                android:id="@+id/bluetooth_audio_channel_mode_dual_channel"
+                layout="@layout/preference_widget_dialog_radiobutton"/>
         </RadioGroup>
 
         <include
diff --git a/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java b/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java
index ea90b4f..ace06d9 100644
--- a/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java
+++ b/src/com/android/settings/development/bluetooth/AbstractBluetoothDialogPreferenceController.java
@@ -57,8 +57,13 @@ public abstract class AbstractBluetoothDialogPreferenceController extends
     protected static final int[] BITS_PER_SAMPLES = {BluetoothCodecConfig.BITS_PER_SAMPLE_32,
             BluetoothCodecConfig.BITS_PER_SAMPLE_24,
             BluetoothCodecConfig.BITS_PER_SAMPLE_16};
+    // CHANNEL_MODE_DUAL_CHANNEL is @hide in BluetoothCodecConfig, so we define it locally
+    // Value must match BluetoothCodecConfig.CHANNEL_MODE_DUAL_CHANNEL (0x1 << 2 = 4)
+    private static final int CHANNEL_MODE_DUAL_CHANNEL = 0x1 << 2;
+
     protected static final int[] CHANNEL_MODES = {BluetoothCodecConfig.CHANNEL_MODE_STEREO,
-            BluetoothCodecConfig.CHANNEL_MODE_MONO};
+            BluetoothCodecConfig.CHANNEL_MODE_MONO,
+            CHANNEL_MODE_DUAL_CHANNEL};
 
     protected final BluetoothA2dpConfigStore mBluetoothA2dpConfigStore;
 
diff --git a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java
index 4a4191a..bc8d454 100644
--- a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java
+++ b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreference.java
@@ -56,14 +56,22 @@ public class BluetoothChannelModeDialogPreference extends BaseBluetoothDialogPre
     }
 
     private void initialize(Context context) {
-        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_default);
-        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_mono);
-        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_stereo);
+        // Load strings first to check array size
         String[] stringArray = context.getResources().getStringArray(
                 com.android.settingslib.R.array.bluetooth_a2dp_codec_channel_mode_titles);
         for (int i = 0; i < stringArray.length; i++) {
             mRadioButtonStrings.add(stringArray[i]);
         }
+
+        // Add radio button IDs - must match string array size
+        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_default);
+        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_mono);
+        mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_stereo);
+        // Only add Dual Channel if the string array has 4 entries (updated translations)
+        if (stringArray.length >= 4) {
+            mRadioButtonIds.add(R.id.bluetooth_audio_channel_mode_dual_channel);
+        }
+
         stringArray = context.getResources().getStringArray(
                 com.android.settingslib.R.array.bluetooth_a2dp_codec_channel_mode_summaries);
         for (int i = 0; i < stringArray.length; i++) {
diff --git a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java
index d156abd..1aa519c 100644
--- a/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java
+++ b/src/com/android/settings/development/bluetooth/BluetoothChannelModeDialogPreferenceController.java
@@ -38,6 +38,10 @@ public class BluetoothChannelModeDialogPreferenceController extends
     private static final String KEY = "bluetooth_channel_mode_settings";
     private static final String TAG = "BtChannelModeCtr";
 
+    // CHANNEL_MODE_DUAL_CHANNEL is @hide in BluetoothCodecConfig, so we define it locally
+    // Value must match BluetoothCodecConfig.CHANNEL_MODE_DUAL_CHANNEL (0x1 << 2 = 4)
+    private static final int CHANNEL_MODE_DUAL_CHANNEL = 0x1 << 2;
+
     public BluetoothChannelModeDialogPreferenceController(Context context, Lifecycle lifecycle,
                                                           BluetoothA2dpConfigStore store) {
         super(context, lifecycle, store);
@@ -57,6 +61,7 @@ public class BluetoothChannelModeDialogPreferenceController extends
     @Override
     protected void writeConfigurationValues(final int index) {
         int channelModeValue = BluetoothCodecConfig.CHANNEL_MODE_NONE; // default
+        long codecSpecific1 = 0; // default
         switch (index) {
             case 0:
                 final BluetoothCodecConfig currentConfig = getCurrentCodecConfig();
@@ -71,10 +76,16 @@ public class BluetoothChannelModeDialogPreferenceController extends
             case 2:
                 channelModeValue = BluetoothCodecConfig.CHANNEL_MODE_STEREO;
                 break;
+            case 3:
+                channelModeValue = CHANNEL_MODE_DUAL_CHANNEL;
+                // Set magic value to enable SBC HD high bitrate mode (551-601 kbps)
+                codecSpecific1 = 0x1337;
+                break;
             default:
                 break;
         }
         mBluetoothA2dpConfigStore.setChannelMode(channelModeValue);
+        mBluetoothA2dpConfigStore.setCodecSpecific1Value(codecSpecific1);
     }
 
     @Override
@@ -98,6 +109,14 @@ public class BluetoothChannelModeDialogPreferenceController extends
                     selectableIndex.add(convertCfgToBtnIndex(CHANNEL_MODES[i]));
                 }
             }
+            // Force-enable Dual Channel for SBC codec - it's always supported by the SBC spec,
+            // even if headphones don't advertise it in their capabilities bitmask.
+            if (currentConfig.getCodecType() == BluetoothCodecConfig.SOURCE_CODEC_TYPE_SBC) {
+                int dualChannelIndex = convertCfgToBtnIndex(CHANNEL_MODE_DUAL_CHANNEL);
+                if (!selectableIndex.contains(dualChannelIndex)) {
+                    selectableIndex.add(dualChannelIndex);
+                }
+            }
         }
         return selectableIndex;
     }
@@ -112,6 +131,9 @@ public class BluetoothChannelModeDialogPreferenceController extends
             case BluetoothCodecConfig.CHANNEL_MODE_STEREO:
                 index = 2;
                 break;
+            case CHANNEL_MODE_DUAL_CHANNEL:
+                index = 3;
+                break;
             default:
                 Log.e(TAG, "Unsupported config:" + config);
                 break;
-- 
2.54.0

